allow creation of pg-stac secrets from azure secret vault, refs #186

batpad opened the pull request:
Refs #186 

Creates the `pgstac-secret-` object reading from an Azure Secrets Vault.

We may or may not want to include this in a main `eoapi-k8s` release, but it would definitely be helpful to have a helm chart with this to test our deploy.

cc @geohacker @emmanuelmathot 

[EDIT] Emmanuel 08/04/2025

- [X] Refactored specific azure service account to generic service account management
- [X] Added extra placeholders in values.yaml to mount extra volumes and extra secret managed by azure in a generic way
- [X] Deactivate existing postgres secret when `postgrescluster.enabled` is false
- [X] Added Azure specific documentation for using Azure managed postgres and keyvault
- [X] Added pod debug logs when startup fails in CI/CD
- [X] updated PGO to 5.7.0

batpad committed:
allow creation of pg-stac secrets from azure secret vault, refs #186

batpad committed:
bump version in chart

batpad committed:
add serviceAccount to create azure service account from values

batpad committed:
change version in chart yaml

batpad committed:
fix secret order since we need multiple secrets derived from a single key in the vault

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-3 and add support for Azure AKS secrets provider

emmanuelmathot committed:
refactor deployment templates to properly handle Azure AKS secrets provider configuration

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-4

emmanuelmathot committed:
fix: update template to use correct context for Azure AKS secrets provider

emmanuelmathot committed:
fix: adjust secret reference formatting and add service account configuration for Azure AKS

emmanuelmathot committed:
fix: add volume mounts for Azure AKS secrets provider in deployment template

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-7

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-8 and add Azure AKS secrets provider configuration

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-9 and add Azure workload identity annotation in deployment template

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-10 and add service account configuration in deployment template

geohacker committed:
use the service account name from values

geohacker committed:
name in the deploy for service account should match values

batpad committed:
key names in vault still not working, try replacing _ with -

batpad committed:
bump version in Chart.yaml

batpad committed:
oops, replace _ with - everywhere

batpad committed:
bump chart version to azure-test-14

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-15 and update job.yaml for Azure workload identity and secrets management

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-16 and update job.yaml for conditional secret inclusion

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-17 and update job.yaml for environment variable configuration

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-18 and update job.yaml for PostgreSQL connection string format

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-19 and update job.yaml and configmap.yaml for PGADMIN_URI initialization and PostgreSQL environment variables

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-20 and update configmap.yaml to ensure PGADMIN_URI is initialized correctly

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-21 and update PGADMIN_URI initialization in configmap.yaml

emmanuelmathot committed:
bump chart version to 0.5.3-azure-test-21 and update job.yaml and configmap.yaml for PGADMIN_URI initialization and PostgreSQL environment variables

emmanuelmathot committed:
Merge branch 'feature/azure-secrets' of https://github.com/developmentseed/eoapi-k8s into feature/azure-secrets

batpad commented:
It looks like we will continue to need something like this to run `eoapi-k8s` for IFRC on Azure.

Wondering how we can clean-up this branch so that we feel it is mergeable.

Am going to mull over this, but just pinging here to get thoughts - @emmanuelmathot @sunu - do one of you feel like you have a clear-ish idea on what it would take to make this code mergeable? And / or any thoughts and perspectives here on the path forward? Thanks :)

emmanuelmathot committed:
update deployment.yaml to allow forwarded IPs in root path configuration

emmanuelmathot committed:
Refactor Azure integration: deprecate specific service account configuration, update pgstacBootstrap settings, and enhance documentation for Azure setup

emmanuelmathot committed:
Add support for additional labels, environment variables, and volume mounts in raster, multidim, stac, and vector configurations

emmanuelmathot commented on docs/azure.md:
```diff
@@ -0,0 +1,234 @@
+# Microsoft Azure Setup
+
+## Using Azure managed PostgreSQL
+To use Azure managed PostgreSQL with the `eoapi-k8s` chart, you need to set up the following:
+1. **Create an Azure PostgreSQL server**: You can create a PostgreSQL server using the Azure portal or the Azure CLI. Make sure to note down the server name, username, and password.
+2. **Create a PostgreSQL database**: After creating the server, create a database that will be used by the `eoapi-k8s` chart.
+3. **Configure firewall rules**: Ensure that the PostgreSQL server allows connections from your Kubernetes cluster's IP address. You can do this by adding a firewall rule in the Azure portal or using the Azure CLI.
```
@geohacker I need your review on this part or a link to a reference documentation to setup an azure managed postgres with postgis and the related keyvault creation and access

emmanuelmathot committed:
Remove redundant value assignment for KEEP_ALIVE in pgstacbootstrap job configuration

emmanuelmathot committed:
Enable backups in PostgreSQL operator installation

emmanuelmathot committed:
Enable backups for PostgreSQL cluster in test configuration

emmanuelmathot committed:
Enhance cleanup step to extract and display pod logs for debugging on failure

emmanuelmathot committed:
Enhance cleanup step to extract and display logs from PGSTACBootstrap and raster pod containers for improved debugging

emmanuelmathot committed:
Refactor ServiceAccount name in RBAC configuration to use template function for consistency

emmanuelmathot committed:
Refactor endpoint exports in helm-tests workflow for consistency and clarity

emmanuelmathot committed:
Fix endpoint URLs in helm-tests workflow to include release name for proper routing

emmanuelmathot committed:
Enhance error handling in test execution by extracting and displaying pod logs for improved debugging

sunu commented on helm-chart/eoapi/test-k3s-unittest-values.yaml:
```diff
@@ -3,6 +3,10 @@ testing: true
+postgrescluster:
+  enabled: true
+  # Needed with PGO <= 5.7.0 https://github.com/CrunchyData/postgres-operator/issues/3152
+  backupsEnabled: true
```
I think we should upgrade to PGO >= 5.7.0 in the tests instead.

emmanuelmathot replied:
Yes indeed but I wanted to avoid testing the new release and to that later

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -401,6 +463,25 @@ vector:
 docServer:
   enabled: true
+
+# TODO DEPRECATE THIS SECTION
```
I would prefer we create a separate deployment chart containing this. Could even live in the same repo. Perhaps `helm-chart/eoapi-azure`. But it seems we add unnecessary tech debt with with cloud provider specific code.

emmanuelmathot committed:
remove azure specifics

emmanuelmathot committed:
Update PGO_VERSION to 5.7.0 and remove backupsEnabled comment from test values

emmanuelmathot committed:
Revert chart version to 0.5.3-azure-test-21 for consistency with application versioning

emmanuelmathot committed:
Update chart version to 0.5.3-azure-test-21 for consistency with application versioning

emmanuelmathot committed:
Add service account configuration and update values.yaml with ingress and environment settings

pantierra commented on .github/workflows/helm-tests.yml:
```diff
@@ -9,7 +9,7 @@ on:
 env:
   HELM_VERSION: v3.15.2
-  PGO_VERSION: 5.5.2
+  PGO_VERSION: 5.7.0
```
I would suggest to align with `5.7.4` which we are using currently. This will also make it easier not to miss when we upgrade again.

emmanuelmathot committed:
Update database connection wait command to use POSTGRES_HOST variable

emmanuelmathot committed:
Add environment variables to wait-for-db init container

emmanuelmathot committed:
Enable pgstacBootstrap and update environment variables for database connection

emmanuelmathot committed:
Fix wait-for-db command to correctly reference POSTGRES_PORT environment variable

emmanuelmathot committed:
Enhance pgstacbootstrap job configuration by adding extra volume mounts and environment variables for LOAD_FIXTURES and KEEP_ALIVE

emmanuelmathot committed:
Refactor pgstacBootstrap to support additional environment variables and update values.yaml to use a map for extraEnvVars

emmanuelmathot committed:
Refactor deployment.yaml to correctly reference extraEnvVars in pgstacBootstrap settings and clean up values.yaml by removing unused envVars for raster and vector

emmanuelmathot committed:
Refactor pgstacBootstrap job configuration to use extraEnvFrom and extraEnvVars for environment variable management in job.yaml and values.yaml

emmanuelmathot committed:
Fix reference to extraEnvFrom in pgstacBootstrap job configuration

emmanuelmathot committed:
Refactor pgstacBootstrap configuration to add extraEnvFrom, extraVolumeMounts, and extraVolumes for improved secret management and volume handling

emmanuelmathot committed:
Fix DUMMY_ENV_VAR value type in deployment.yaml to use string format

emmanuelmathot committed:
Refactor PostgreSQL configuration and remove deprecated database setup

- Introduced a unified PostgreSQL configuration structure in values.yaml, replacing the old db configuration.
- Added new helper functions for managing PostgreSQL environment variables and secrets based on the selected configuration type (postgrescluster, external-plaintext, external-secret).
- Removed old database-related templates (ConfigMap, Deployment, PVC, Secrets, Service) that are no longer needed.
- Updated the pgstacbootstrap job and configmap templates to align with the new PostgreSQL configuration.
- Implemented validation for PostgreSQL settings to ensure required fields are set based on the selected type.

emmanuelmathot committed:
Add PostgreSQL host reader and writer environment variables, and include DATABASE_URL for connection string

emmanuelmathot committed:
Merge remote-tracking branch 'origin/main' into unified_pg

emmanuelmathot committed:
Merge branch 'main' into unified_pg

emmanuelmathot committed:
Added a clarifying comment in values.yaml to explain that values in the external secret (host, port, database) will override the corresponding values defined in external.host, external.port, and external.database.

Confirmed that the conditional blocks in deployment.yaml were already consolidated to eliminate redundancy. The file was already using a single include statement for PostgreSQL environment variables:

env:
  {{- include "eoapi.postgresqlEnv" $ | nindent 12 }}
Removed the unused eoapi.mapLegacyPostgresql helper function from _helpers.tpl as it wasn't being referenced anywhere in the codebase.

emmanuelmathot committed:
Merge branch 'unified_pg' into feature/azure-secrets

emmanuelmathot committed:
Enhance Azure PostgreSQL setup documentation with detailed setup instructions and examples for server creation, database setup, firewall configuration, and Key Vault integration.

emmanuelmathot commented:
need #215 to be merged and closed

emmanuelmathot committed:
Remove unused ingress and values.yaml configurations

emmanuelmathot commented:
@pantierra No more azure specific configuration

emmanuelmathot committed:
Merge branch 'main' into feature/azure-secrets

emmanuelmathot committed:
Bump chart version to 0.6.0 for release
